### PR TITLE
Introducing `vconfig`, and (starting to) tidy up the use of options for verification

### PR DESCRIPTION
**Pull request by mtzguido**
This PR has two parts:

1) Introduce a `vconfig` type with the state of verification options. We use it to implement `sigelt_opts` and `check_with` in Meta-F* without having to embed a full optionstate (which was anyway dangerous, as the recent bug with `--verify_module` showed). This has the added benefit that metaprograms can now read and tweak the options instead of just having an opaque object.

2) Remove all calls to `Options.lax()` in the typechecker, and similarly for `Options.admit_smt_queries()` and `Options.should_verify`. These are now only inspected in `FStar.Universal`, which sets `env.lax` and `env.admit` accordingly. All posterior phases rely only on these fields. One exception is in processing an options pragma, but see future points. This should not be observable at all to users, it's an internal cleanup.

(1) also has the benefit that it makes the size of stored options a lot smaller, so we could just always enable it (instead of conditionally with `--record_options`).

As a next step, I would like to embed `vconfig` inside the typechecker `env`, and remove all calls to the `Options` module for these knobs. The options would remain but only as a default to fill the initial vconfig. This would allow to remove `get_vconfig` and `set_vconfig` in the options module and make it purely a typechecker thing. Ideally, I would prefer if we were to move away from option setting pragmas into some kind of dedicated syntax for modifying the vconfig, to make `Options` less relevant in the later phases such as the encoding.

(This got a yellow for StreamAE, but I think it's just noise.)

### CODE REVIEW

File: src/basic/FStar.VConfig.fs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#light "off"
```
Comment (nikswamy) [inline src/basic/FStar.VConfig.fs:1]: copyright notice

File: src/syntax/FStar.Syntax.Embeddings.fs [hunk 2]
```diff
@@ -708,6 +711,142 @@ let e_range =
         Range.string_of_range
         (ET_app (PC.range_lid |> Ident.string_of_lid, []))
 
+let e_vconfig =
```
Comment (nikswamy) [inline src/syntax/FStar.Syntax.Embeddings.fs:714]: Is it really worth embedding and unembedding vconfig back and forth?

Why not expose a vconfig to metaprograms as an abstract type, with operations to get and set specific options working by manipulating the native object?

Comment (nikswamy) [reply src/syntax/FStar.Syntax.Embeddings.fs:714]: This may also have the benefit of providing an interface to vconfigs for use in NBE

**Comment by nikswamy** [comment]
About this comment:

> Ideally, I would prefer if we were to move away from option setting pragmas into some kind of dedicated syntax for modifying the vconfig, to make Options less relevant in the later phases such as the encoding.

That would be quite a significant change, wouldn't it? We use #push-options etc. pretty heavily in many developments. Can you say more about why you would prefer having a separate syntax for verification configuration options?

File: src/syntax/FStar.Syntax.Embeddings.fs (hunk 2 above)
Comment (mtzguido) [reply src/syntax/FStar.Syntax.Embeddings.fs:714]: The main advantage in my mind is that one can then use record syntax (projection, override, etc) and create one from scratch. But yes, the embeddings are a pain to write. Some times I considered having a small language to autogenerate these embeddings, maybe it's time to bite that bullet.

**Comment by mtzguido** [comment]
Thanks for reviewing Nik!

> That would be quite a significant change, wouldn't it? We use #push-options etc. pretty heavily in many developments. Can you say more about why you would prefer having a separate syntax for verification configuration options?

The different syntax is more of an afterthought. I would like to properly separate (internally) the verification configuration from the command line options. Having a single mutable map with the options, and having to push/pop is pretty complicated for handling (for instance) multiple goals in tactics (each of which currently embeds a whole proofstate to make this work). The implementation of `set_options` for tactics is complicated by this. So I was thinking that if they really separate objects, that could be made clearer from the syntax already. On the other hand, it would allow for top-level "commands" like:

```fstar
let f = ...

set rlimit = rlimit*2

let y = ...
```

Which only affect the vcfg.

**Comment by mtzguido** [comment]
Merging this, except for the change in behaviour of options--- I'll make a separate PR for that so we discuss better.

**Pull request merged**